### PR TITLE
Shadowserver 202207

### PR DESCRIPTION
**Pull request by elsif2**
Large set of changes previously submitted as 2143.

### CODE REVIEW

File: docs/user/bots.rst [hunk 1]
```diff
@@ -654,9 +654,10 @@ The Cache is required to memorize which files have already been processed (TTL n
 
 **Configuration Parameters**
 
-* `country`: The country you want to download the reports for
+* `country`: **Deprecated:** The country you want to download the reports for. Will be removed in IntelMQ version 4.0.0, use *reports* instead.
 * `apikey`: Your Shadowserver API key
 * `secret`: Your Shadowserver API secret
+* `reports`: A list of strings or a comma-separated list of the mailing lists you want to process.
```
Comment (sebix) [inline docs/user/bots.rst:660]: What is a "mailing list" in this context?

File: intelmq/bots/collectors/shadowserver/collector_reports_api.py [hunk 2]
```diff
@@ -22,12 +23,14 @@
...
 class ShadowServerAPICollectorBot(CollectorBot, HttpMixin, CacheMixin):
     """
-    Connects to the Shadowserver API, requests a list of all the reports for a specific country and processes the ones that are new
+    Connects to the Shadowserver API, requests a list of all the reports for an organization and processes the ones that are new
 
     Parameters:
         api_key (str): Your Shadowserver API key
         secret (str): Your Shadowserver API secret
-        country (str): The country you want to download reports for (i.e. 'austria')
+        country (str): DEPRECIATED The mailing list you want to download reports for (i.e. 'austria')
```
Comment (sebix) [inline intelmq/bots/collectors/shadowserver/collector_reports_api.py:31]:
```suggestion
        country (str): DEPRECATED The mailing list you want to download reports for (i.e. 'austria')
```

File: intelmq/bots/collectors/shadowserver/README.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<!--
+SPDX-FileCopyrightText: 2022 The Shadowserver Foundation
+SPDX-License-Identifier: AGPL-3.0-or-later
+-->
+
+This module is maintained by [The Shadowserver Foundation](https://www.shadowserver.org/).  
+
+Please contact intelmq@shadowserver.org with any issues or concerns.
+
```
Comment (sebix) [inline intelmq/bots/collectors/shadowserver/README.md]:
```suggestion
```

File: intelmq/bots/parsers/shadowserver/README.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<!--
+SPDX-FileCopyrightText: 2022 The Shadowserver Foundation
+SPDX-License-Identifier: AGPL-3.0-or-later
+-->
+
+This module is maintained by [The Shadowserver Foundation](https://www.shadowserver.org/).  
+
+Please contact intelmq@shadowserver.org with any issues or concerns.
+
```
Comment (sebix) [inline intelmq/bots/parsers/shadowserver/README.md:9]:
```suggestion
```

File: intelmq/lib/test.py [hunk 5]
```diff
@@ -46,7 +46,7 @@
               }
 
 
-class Parameters:
+class Parameters(object):
```
Comment (sebix) [inline intelmq/lib/test.py]:
```suggestion
class Parameters():
```
Adding object as parent is not necessary on Python 3 any more

File: intelmq/lib/test.py [hunk 6]
```diff
@@ -63,7 +63,7 @@ def mocked(conf_file):
             confname = os.path.join('etc/', os.path.split(conf_file)[-1])
             fname = pkg_resources.resource_filename('intelmq',
                                                     confname)
-            with open(fname) as fpconfig:
+            with open(fname, 'rt') as fpconfig:
```
Comment (sebix) [inline intelmq/lib/test.py]: why is this change needed here?

File: intelmq/lib/test.py [hunk 7]
```diff
@@ -105,7 +105,7 @@ def skip_build_environment():
     return unittest.skipIf(os.getenv('USER') == 'abuild', 'Test disabled in Build Service.')
 
 
-class BotTestCase:
+class BotTestCase(object):
```
Comment (sebix) [inline intelmq/lib/test.py]:
```suggestion
class BotTestCase():
```

File: intelmq/lib/test.py [hunk 8]
```diff
@@ -199,11 +199,12 @@ def prepare_bot(self, parameters={}, destination_queues=None):
...
 
-        src_name = f"{self.bot_id}-input"
+        src_name = "{}-input".format(self.bot_id)
         if not destination_queues:
-            destination_queues = {"_default": f"{self.bot_id}-output"}
+            destination_queues = {"_default": "{}-output".format(self.bot_id)}
         else:
-            destination_queues = {queue_name: f"{self.bot_id}-{queue_name.strip('_')}-output"
+            destination_queues = {queue_name: "%s-%s-output" % (self.bot_id,
+                                                                queue_name.strip('_'))
```
Comment (sebix) [inline intelmq/lib/test.py]: I don't see a relation to the shadowserver changes

File: intelmq/tests/bots/parsers/shadowserver/README.md [hunk 9]
```diff
@@ -14,3 +15,8 @@ each type of report, there are also the following tests:
 * `test_broken.py`: to test errors in the parser
 * `test_testdata.py`: to test if the parsing of large original files
 * `test_parameters.py`: to test the bot's parameters
+
+This module is maintained by [The Shadowserver Foundation](https://www.shadowserver.org/).  
+
+Please contact intelmq@shadowserver.org with any issues or concerns.
+
```
Comment (sebix) [inline intelmq/tests/bots/parsers/shadowserver/README.md]:
```suggestion
```

**Review by sebix — COMMENTED** [review]
Would appreciate if you could squelch the commits together (not necessarily to one, but < 10)

File: docs/user/bots.rst (hunk 1 above)
Comment (elsif2) [reply docs/user/bots.rst:660]: Mailing lists are how reports were delivered before API delivery was available.  Some organizations had multiple mailing lists based on a filter type (e.g. geo and IP/CIDR).

File: intelmq/lib/test.py (hunk 5 above)
Comment (elsif2) [reply intelmq/lib/test.py]: I don't believe I modified intelmq/lib/test.py - this must be a synchronization issue

**Comment by elsif2** [comment]
I squashed the changes into a small number of commits.

File: CHANGELOG.md [hunk 10]
```diff
@@ -100,6 +100,8 @@ CHANGELOG
 - `intelmq.bots.outputs.templated_smtp.output`:
   - Add new function `from_json()` (which just calls `json.loads()` in the standard Python environment), meaning the Templated SMTP output bot can take strings containing JSON documents and do the formatting itself (PR#2120 by Karl-Johan Karlsson).
   - Lift restriction on requirement jinja2 < 3 (PR#2158 by Sebastian Wagner).
+- `intelmq.bots.outputs.sql`:
+  - For PostgreSQL, escape Nullbytes in text to prevent "unsupported Unicode escape sequence" issues (PR#2223 by Sebastian Wagner, fixes #2203).
```
Comment (sebix) [inline CHANGELOG.md]: This change shouldn't be here

File: intelmq/lib/test.py [hunk 11]
```diff
@@ -188,7 +188,7 @@ def log(name, *args, **kwargs):
             return logger
         return log
 
-    def prepare_bot(self, parameters={}, destination_queues=None):
+    def prepare_bot(self, parameters={}, destination_queues=None, prepare_source_queue: bool = True):
```
Comment (sebix) [inline intelmq/lib/test.py]: There are also still changes in unrelated parts of the code

**Comment by sebix** [comment]
> I squashed the changes into a small number of commits.

This PR has 95 commits?

**Comment by elsif2** [comment]
> > I squashed the changes into a small number of commits.
> This PR has 95 commits?

I did a `git rebase -i develop` and squashed down to three which required a lot of merge repairs.

The git log shows them as follows:

```
commit 67a52fe8e8d4c965970af8884ae627e272d9002c
Author: elsif2 <elsif@shadowserver.org>
Date:   Wed Jan 26 17:02:51 2022 +0000

    parent 7dc5b7471b952d4fd703c2d8921ea992296194f3
    author elsif2 <elsif@shadowserver.org> 1643216571 +0000
    committer elsif2 <elsif@shadowserver.org> 1659395284 +0000
    
    parent 7dc5b7471b952d4fd703c2d8921ea992296194f3
    author elsif2 <elsif@shadowserver.org> 1643216571 +0000
    committer elsif2 <elsif@shadowserver.org> 1659395281 +0000
    
    parent 7dc5b7471b952d4fd703c2d8921ea992296194f3
    author elsif2 <elsif@shadowserver.org> 1643216571 +0000
    committer elsif2 <elsif@shadowserver.org> 1659395278 +0000
    
    parent 7dc5b7471b952d4fd703c2d8921ea992296194f3
    author elsif2 <elsif@shadowserver.org> 1643216571 +0000
    committer elsif2 <elsif@shadowserver.org> 1659395264 +0000
 .
 .
 .
```

**Comment by sebix** [comment]
GitHub says 95 commits
image
command line says 93:
```
$ git log --oneline  upstream/pr/2227...develop | wc -l
93
```

File: intelmq/lib/test.py (hunk 11 above)
Comment (elsif2) [reply intelmq/lib/test.py]: I don't see any changes compared to the develop branch:

```
$ git status
On branch shadowserver-202207
Your branch is up to date with 'origin/shadowserver-202207'.

nothing to commit, working tree clean
```

```
$ curl -s 'https://raw.githubusercontent.com/certtools/intelmq/develop/intelmq/lib/test.py' > /tmp/test.py
$ diff -u intelmq/lib/test.py /tmp/test.py
$
```

**Comment by elsif2** [comment]
What would you suggest?

**Comment by sebix** [comment]
```
git rebase -i develop
```
then squash various commit together and
```
git push -f origin shadowserver-202207
```

**Comment by elsif2** [comment]
> ```
> git rebase -i develop
> ```
> 
> then squash various commit together and
> 
> ```
> git push -f origin shadowserver-202207
> ```

Done:

> [elsif2](https://github.com/elsif2) wants to merge 3 commits into [develop](https://github.com/certtools/intelmq) from [shadowserver-202207](https://github.com/certtools/intelmq/tree/shadowserver-202207)

**Comment by sebix** [comment]
Cool, looks tidy now :)

**Comment by aaronkaplan** [comment]
Agreement is that I merge it in, if things break, @elsif2 will fix it quickly. 
IMHO it is very important to have the new shadowserver code in 3.1. However, please note that we should really test this in a sample environment before tagging the release. Agreed, @sebix ?

Thanks @elsif2 for your enormous heavy lifting and improvements!
Can't over-state how valuable this is.
